crates/wasmtime/src/component/values.rs

```diff
@@ -0,0 +1,389 @@
+            }
+            Val::U8(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(*value as u32));
+            }
+            Val::S16(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i32(*value as i32));
+            }
+            Val::U16(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(*value as u32));
+            }
+            Val::S32(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i32(*value));
+            }
+            Val::U32(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(*value));
+            }
+            Val::S64(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i64(*value));
+            }
+            Val::U64(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u64(*value));
+            }
+            Val::Float32(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::f32(*value));
+            }
+            Val::Float64(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::f64(*value));
+            }
+            Val::Char(value) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(u32::from(*value)));
+            }
+            Val::String(value) => {
+                let (ptr, len) = super::lower_string(
+                    &mut MemoryMut::new(store.as_context_mut(), options),
+                    value,
+                )?;
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i64(ptr as i64));
+                map_maybe_uninit!(dst[offset + 1]).write(ValRaw::i64(len as i64));
+            }
+            Val::List(List { values, .. }) => {
+                let (ptr, len) =
+                    lower_list(&mut MemoryMut::new(store.as_context_mut(), options), values)?;
+                map_maybe_uninit!(dst[offset]).write(ValRaw::i64(ptr as i64));
+                map_maybe_uninit!(dst[offset + 1]).write(ValRaw::i64(len as i64));
+            }
+            Val::Record(Record { values, .. }) | Val::Tuple(Tuple { values, .. }) => {
+                for (index, value) in values.iter().enumerate() {
+                    value.lower(store, options, dst, offset + index)?;
+                }
+            }
+            Val::Variant(Variant {
+                discriminant,
+                value,
+                ..
+            })
+            | Val::Union(Union {
+                discriminant,
+                value,
+                ..
+            })
+            | Val::Option(Option {
+                discriminant,
+                value,
+                ..
+            })
+            | Val::Expected(Expected {
+                discriminant,
+                value,
+                ..
+            }) => {
+                map_maybe_uninit!(dst[offset]).write(ValRaw::u32(*discriminant));
+                value.lower(store, options, dst, offset + 1)?;
```

Comment:
This is another case that I think is not quite right for subtle reasons. The main one is that unused `ValRaw` entries (and upper bits which are unused here) need to all be set to zero. Otherwise this runs the risk of leaking undefined values into wasm modules.

Additionally this is another location that needs to inform the caller that a nonzero amount of `ValRaw` slots are consumed. Given `value` here you may also consume more slots than `value` takes up because of how variants are "flattened"